tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -138,34 +141,67 @@ __find_best_conn(TfwSrvConn **best_conn, TfwHashSrv *srv_cl,
  *  - For every HTTP request, we have to scan the list of all servers to find
  *    a matching one with the highest weight. That adds some overhead.
  */
+static inline TfwSrvConn *
+__find_best_conn(TfwMsg *msg, TfwHashConnList *cl)
+{
+	ssize_t l_idx, r_idx, idx;
+	TfwSrvConn *conn;
+	unsigned long msg_hash = tfw_http_req_key_calc((TfwHttpReq *)msg);
+	unsigned long best_hash = (~0UL ^ msg_hash);
+
+	if (unlikely(!cl->conn_n))
+		return NULL;
+
+	/*
+	 * Find a connection with hash as close to @best_hash as possible.
+	 * Value of (msg_hash ^ srv_conn_hash) will be biggest for that
+	 * connection.
+	 */
+	idx = __bsearch(best_hash, cl);
+	conn = cl->conns[idx].conn;
+	if (likely(!tfw_srv_conn_restricted(conn)
+		   && !tfw_srv_conn_queue_full(conn)
+		   && tfw_srv_conn_get_if_live(conn)))
+		return conn;
+
+	/*
+	 * The best connection is dead or overfilled. Take the nearest live
+	 * neighbour.
+	*/
+	r_idx = idx + 1;
+	l_idx = idx - 1;
+	while (l_idx >= 0 || r_idx < (ssize_t)cl->conn_n) {
+		unsigned long l_diff = (l_idx >= 0)
+				? (best_hash - cl->conns[l_idx].hash)
+				: ULONG_MAX;
+		unsigned long r_diff = (l_idx >= 0)
+				? (cl->conns[r_idx].hash - best_hash)
+				: ULONG_MAX;
+		ssize_t best_idx = (l_diff <= r_diff) ? l_idx : r_idx;
+
+		conn = cl->conns[best_idx].conn;
+		if (likely(!tfw_srv_conn_restricted(conn)
+			   && !tfw_srv_conn_queue_full(conn)
+			   && tfw_srv_conn_get_if_live(conn)))
```

Comment:
This condition is repeated twice. Perhaps, it makes sense to make an inline function with it? Like `tfw_sched_hash_conn_suitable()`?